File: 2015/day02.exs
defmodule AOC2015.Day2 do
  def import(file) do
    {:ok, content} = File.read(file)

    content
    |> String.split("\n", trim: true)
    |> Enum.map(&convert_to_data/1)
  end

  defp convert_to_data(input) do
    input
    |> String.split("x")
    |> Enum.map(&String.to_integer/1)
    |> Enum.sort()
    |> List.to_tuple()
  end

  defp needed_paper({a, b, c}) do
    3 * a * b + 2 * a * c + 2 * b * c
  end

  defp needed_ribbon({a, b, c}) do
    2 * a + 2 * b + a * b * c
  end

  def task1(input) do
    input
    |> Enum.map(&needed_paper/1)
    |> Enum.sum()
  end

  def task2(input) do
    input
    |> Enum.map(&needed_ribbon/1)
    |> Enum.sum()
  end
end

input = AOC2015.Day2.import("input_day02.txt")

input
|> AOC2015.Day2.task1()
|> IO.puts()

input
|> AOC2015.Day2.task2()
|> IO.puts()

File: ruby/2015/day02.rb
file = File.open("../../input/2015/input_day02.txt", "r")

area = 0
ribbon = 0
while line = file.gets
  line = line.split('x').map(&:to_i).sort
  area += 2*line[0]*line[1] + 2*line[0]*line[2] + 2*line[1]*line[2] + line[0]*line[1]
  ribbon += 2*line[0] + 2*line[1] + line[0]*line[1]*line[2]
end
file.close

puts area
puts ribbon
